Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -2783,6 +2882,23 @@ def create_class_from_scope(self, node, target_module_scope, inner_node=None):
             if entry.is_declared_generic:
                 closure_entry.is_declared_generic = 1
         node.needs_closure = True
+
+        ninputs = 0
+        for var in inputs_in_closure:
```

Comment:
Are these really so special that they merit their own declaration loop? Can't they be integrated with the normal (temp) variable handling?

What you could try is to use a `TempNode` for the references. I've never seen those being used across closure boundaries, so there might be obstacles (which could genuinely be worth removing), but they seem an appropriate mechanism for handling this kind of temporary expression extraction.

Reply:
The main issue I had with the normal temp handling was that temps are usually put in `closure_scope->__pyx_tx` rather than `closure_scope->outer_scope->__pyx_tx`, and we don't have access to `closure_scope` in the function doing the setup.

However I agree with your dislike of special cases, so I'll have another think about this when I come back to it.